lib/puppet/functions/stdlib/sort_by.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+# @summary Sort an Array, Hash or String by mapping values through a given block.
+#
+# @example Sort local devices according to their used space.
+#   $facts['mountpoints'].stdlib::sort_by |$m| { $m.dig(1, 'used_bytes') }
+#
+Puppet::Functions.create_function(:'stdlib::sort_by') do
+  # @param ary The Array to sort.
+  # @param block The block for transforming elements of ary.
+  # @return [Array] Returns an ordered copy of ary.
+  dispatch :sort_by_array do
+    param 'Array', :ary
+    block_param 'Callable[1,1]', :block
+  end
+
+  # @param str The String to sort.
+  # @param block The block for transforming elements of str.
+  # @return [String] Returns an ordered copy of str.
+  dispatch :sort_by_string do
+    param 'String', :str
+    block_param 'Callable[1,1]', :block
+  end
+
+  # @param hsh The Hash to sort.
+  # @param block The block for transforming elements of hsh.
+  # @return [Hash] Returns an ordered copy of hsh.
+  dispatch :sort_by_hash1 do
+    param 'Hash', :hsh
+    block_param 'Callable[1,1]', :block
+  end
+
+  # @param hsh The Hash to sort.
+  # @param block The block for transforming elements of hsh.
+  # @return [Hash] Returns an ordered copy of hsh.
+  dispatch :sort_by_hash2 do
+    param 'Hash', :hsh
+    block_param 'Callable[2,2]', :block
+  end
+
+  def sort_by_iterable(iterable, &block)
+    Puppet::Pops::Types::Iterable.asserted_iterable(self, iterable).sort_by(&block)
+  end
+
+  def sort_by_array(ary, &block)
+    sort_by_iterable(ary, &block)
+  end
+
+  def sort_by_string(str, &block)
+    sort_by_iterable(str, &block).join
+  end
+
+  def sort_by_hash1(hsh, &block)
+    sort_by_iterable(hsh, &block).to_h
+  end
+
+  def sort_by_hash2(hsh)
+    sort_by_iterable(hsh) { |e| yield(*e) }.to_h
```

Comment:
`e` makes me think, I think this is more readable with explicit names:
```suggestion
    sort_by_iterable(hsh) { |key, value| yield(key, value) }.to_h
```

Reply:
`sort_by_hash2` is written this way, because Ruby's `sort_by` expects a block that has just one parameter. I thought, this construct was needed to align `stdlib::sort_by($hash) |$key, $value| { ... }` with that. Though it looks like I was mistaken and Ruby handles that transparently:
```ruby
hsh = { 'The' => 'quick', 'brown' => 'fox', 'jumps' => 'over', 'the' => 'lazy', 'dog' => '.' }
hsh.sort_by { |e| e[1] } # => [["dog", "."], ["brown", "fox"], ["the", "lazy"], ["jumps", "over"], ["The", "quick"]]
hsh.sort_by { |k, v| v } # => [["dog", "."], ["brown", "fox"], ["the", "lazy"], ["jumps", "over"], ["The", "quick"]]
```
My delusion was reinforced by `sort_by()` accepting blocks with arity 2 on Arrays and _not_ interpreting that as index-element-tuple, but rather element-nil-tuple. In other words, `sort_by()` does _not_ use `Enumerable.each_with_index` when appropriate.

Reply:
On that note, [Puppet's `each()`](https://github.com/puppetlabs/puppet/blob/8.0.1_release/lib/puppet/functions/each.rb#L153) explicitly calls Ruby's `each_with_index()` in this case.

Reply:
With the collapsed `sort_by_hash()`, this discussion can be resolved, too.